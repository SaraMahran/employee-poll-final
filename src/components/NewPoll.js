import React, { useState } from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { handleAddQuestion } from "../actions/questions";

function NewPoll({ dispatch, authedUser }) {
  const [optionOneText, setOptionOneText] = useState("");
  const [optionTwoText, setOptionTwoText] = useState("");

  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();

    if (optionOneText && optionTwoText && authedUser) {
      dispatch(
        handleAddQuestion({
          optionOneText,
          optionTwoText,
          author: authedUser,
        })
      );

      setOptionOneText("");
      setOptionTwoText("");

      navigate("/");
    } else {
      console.error("Please provide optionOneText, optionTwoText, and author");
    }
  };

  return (
    <div className="new-poll">
      <h2 className="dark:text-teal-800 text-3xl">Create New Poll</h2>
      <br />
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="optionOne" className="dark:text-teal-500 text-xl">
            Option One
          </label>
          <br />
          <input
            type="text"
            id="optionOne"
            value={optionOneText}
            onChange={(e) => setOptionOneText(e.target.value)}
            placeholder="Enter first option"
          />
        </div>

        <div>
          <label htmlFor="optionTwo" className="dark:text-teal-500 text-xl">
            {" "}
            Option Two
          </label>
          <br />
          <input
            type="text"
            id="optionTwo"
            value={optionTwoText}
            onChange={(e) => setOptionTwoText(e.target.value)}
            placeholder="Enter second option"
          />
        </div>

        <button
          type="submit"
          className="btn btn-primary bg-teal-600  border-2 border-red-200  rounded-lg px-4 py-2 text-red-200"
        >
          Submit Poll
        </button>
      </form>
    </div>
  );
}

const mapStateToProps = ({ authedUser }) => ({
  authedUser,
});

export default connect(mapStateToProps)(NewPoll);
