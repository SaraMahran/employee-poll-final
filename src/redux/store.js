import { configureStore } from "@reduxjs/toolkit";
import authedUser from "../reducers/authedUser";
import users from "../reducers/users";
import questions from "../reducers/questions";

const store = configureStore({
  reducer: {
    authedUser,
    users,
    questions,
  },
});

export default store;
