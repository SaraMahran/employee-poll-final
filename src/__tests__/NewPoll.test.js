import { createStore } from "redux";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewPoll from "../components/NewPoll";
import rootReducer from "../reducers";

const mockStore = createStore(rootReducer, {
  authedUser: "testUser",
});

describe("NewPoll", () => {
  it("should render the component", () => {
    const { asFragment } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <NewPoll />
        </BrowserRouter>
      </Provider>
    );

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display all elements", () => {
    const { getByLabelText, getByText } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <NewPoll />
        </BrowserRouter>
      </Provider>
    );

    expect(getByLabelText("Option One")).toBeInTheDocument();
    expect(getByLabelText("Option Two")).toBeInTheDocument();
    expect(getByText("Submit Poll")).toBeInTheDocument();
  });

  it("should allow users to input text for options", () => {
    const { getByLabelText } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <NewPoll />
        </BrowserRouter>
      </Provider>
    );

    const optionOneInput = getByLabelText("Option One");
    const optionTwoInput = getByLabelText("Option Two");

    fireEvent.change(optionOneInput, { target: { value: "Texas" } });
    fireEvent.change(optionTwoInput, { target: { value: "New Hampshire" } });

    expect(optionOneInput.value).toBe("Texas");
    expect(optionTwoInput.value).toBe("New Hampshire");
  });

  it("should submit the form correctly", () => {
    const mockDispatch = jest.fn();
    const { getByLabelText, getByText } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <NewPoll dispatch={mockDispatch} />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.change(getByLabelText("Option One"), {
      target: { value: "Option 1" },
    });
    fireEvent.change(getByLabelText("Option Two"), {
      target: { value: "Option 2" },
    });

    fireEvent.click(getByText("Submit Poll"));

    expect(mockDispatch).toHaveBeenCalledWith({
      type: "ADD_QUESTION",
      question: {
        optionOneText: "Option 1",
        optionTwoText: "Option 2",
        author: "testUser",
      },
    });
  });
});
