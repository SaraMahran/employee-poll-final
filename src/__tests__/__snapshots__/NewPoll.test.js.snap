// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPoll should render the component 1`] = `
<DocumentFragment>
  <div
    class="new-poll"
  >
    <h2
      class="dark:text-teal-800 text-3xl"
    >
      Create New Poll
    </h2>
    <br />
    <form>
      <div>
        <label
          class="dark:text-teal-500 text-xl"
          for="optionOne"
        >
          Option One
        </label>
        <br />
        <input
          id="optionOne"
          placeholder="Enter first option"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="dark:text-teal-500 text-xl"
          for="optionTwo"
        >
           Option Two
        </label>
        <br />
        <input
          id="optionTwo"
          placeholder="Enter second option"
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary bg-teal-600  border-2 border-red-200  rounded-lg px-4 py-2 text-red-200"
        type="submit"
      >
        Submit Poll
      </button>
    </form>
  </div>
</DocumentFragment>
`;
